roachprod: better determination if scp -R flag can be used

RaduBerinde opened the pull request:
When uploading a file to a cluster, we use the "tree dist" algorithm by default. This uploads the file to a single node, then we copy the file from that node to the other nodes (up to 10).

This only makes sense if the remote-to-remote transfers can happen directly, which only happens if we pass the `-R -A` flags to `scp`. Unfortunately older versions don't support these flags. Currently the flags are only passed if the OS is `darwin`.

This commits improves the determination - we run `ssh -V` (once) and check if the `SSL` major version is three. For reference, some examples of what `ssh -V` returns:
 - recent MacOSX: `OpenSSH_9.0p1, LibreSSL 3.3.6`
 - Ubuntu 22.04: `OpenSSH_8.9p1 Ubuntu-3ubuntu0.3, OpenSSL 3.0.2 15 Mar 2022`

In addition, if the version is not 3, we disable the use of "tree dist".

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109036)
<!-- Reviewable:end -->

stevendanna commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -2439,6 +2440,24 @@ func (c *SyncedCluster) SSH(ctx context.Context, l *logger.Logger, sshArgs, args
+		cmd := exec.Command("ssh", "-V")
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			panic(fmt.Sprintf("error running ssh -V: %v", err))
```
This should be rare, think it's worth logging the output as well?

stevendanna commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -2439,6 +2440,24 @@ func (c *SyncedCluster) SSH(ctx context.Context, l *logger.Logger, sshArgs, args
+		if err != nil {
+			panic(fmt.Sprintf("error running ssh -V: %v", err))
+		}
+		sshVersion3Internal.value = strings.Contains(string(out), "SSL 3.")
```
We could try parsing to version to future this against version 4. But eh, I suppose that is next decades problem.

RaduBerinde replied:
Plus if history is any guide, next version might be 5. Or 9? :)

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109036#-:-NcYX-Kb24Ux4lhraF1y:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @aliher1911, @herkolategan, @smg260, and @stevendanna)

___
*[`pkg/roachprod/install/cluster_synced.go` line 2454 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/109036#gh-1299349066:-NcYX0BTCy7SSWV-etx6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/506552b5e4fb0e7158dcf466ffbd6d821efe4ec4/pkg/roachprod/install/cluster_synced.go#L2454)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

This should be rare, think it's worth logging the output as well?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
roachprod: better determination if scp -R flag can be used

When uploading a file to a cluster, we use the "tree dist" algorithm
by default. This uploads the file to a single node, then we copy the
file from that node to the other nodes (up to 10).

This only makes sense if the remote-to-remote transfers can happen
directly, which only happens if we pass the `-R -A` flags to `scp`.
Unfortunately older versions don't support these flags. Currently the
flags are only passed if the OS is `darwin`.

This commits improves the determination - we run `ssh -V` (once) and
check if the `SSL` major version is three. For reference, some
examples of what `ssh -V` returns:
 - recent MacOSX: `OpenSSH_9.0p1, LibreSSL 3.3.6`
 - Ubuntu 22.04: `OpenSSH_8.9p1 Ubuntu-3ubuntu0.3, OpenSSL 3.0.2 15 Mar 2022`

In addition, if the version is not 3, we disable the use of "tree
dist".

Epic: none
Release note: None

RaduBerinde commented:
bors r+

renatolabs commented on pkg/roachprod/install/cluster_synced.go:
```diff
@@ -1752,7 +1751,9 @@ func (c *SyncedCluster) Put(
-		if c.UseTreeDist {
+		// If scp does not support the -R flag, tree dist is slower as it ends up
+		// transferring everything through this machine anyway.
+		if c.UseTreeDist && sshVersion3() {
 			detail = " (dist)"
```
I'm confused -- this diff is changing the `detail` that is logged when transferring files, but not changing the actual logic of transferring these files. IIUC, this will change message logged on TC nightly builds (where agents  run Ubuntu 20.04) but that's confusing since the transferring algorithm won't have changed.

Am I missing something?

RaduBerinde commented:
bors r-

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109036)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @aliher1911, @herkolategan, @renatolabs, @smg260, and @stevendanna)

___
*[`pkg/roachprod/install/cluster_synced.go` line 1757 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/109036#gh-1303430618:-NcYsW3XAvFfJtDPFBt7:b-9sythc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1f90242d28549d0bace69a0679397be2595768/pkg/roachprod/install/cluster_synced.go#L1757)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I'm confused -- this diff is changing the `detail` that is logged when transferring files, but not changing the actual logic of transferring these files. IIUC, this will change message logged on TC nightly builds (where agents  run Ubuntu 20.04) but that's confusing since the transferring algorithm won't have changed.

Am I missing something?
</blockquote></details>

Ah.. oops, :blush:   Fixed, PTAL


<!-- Sent from Reviewable.io -->

herkolategan reviewed (approved):
Nice!

RaduBerinde commented:
bors r+
